warehouse/query-core/src/main/java/datawave/next/scanner/DocumentRangeScan.java

```diff
@@ -0,0 +1,250 @@
+
+        // set the query from the query data
+        if (setting.getOptions().containsKey(QueryOptions.QUERY)) {
+            setting.addOption(QueryOptions.QUERY, queryData.getQuery());
+        }
+
+        try (Scanner scanner = config.getClient().createScanner(keyWithContext.getContext().getTableName(), auths)) {
+            scanner.addScanIterator(setting);
+
+            Key start = new Key(keyWithContext.getKey().getRow());
+            scanner.setRange(new Range(start, true, start.followingKey(PartialKey.ROW), false));
+
+            driveScanner(scanner);
+
+        } catch (Exception e) {
+            log.error("exception while fetching document", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void copyRequiredOptions(IteratorSetting target, IteratorSetting source) {
+        Set<String> requiredOptions = DocumentIteratorOptions.getRequiredOptionNames();
+        Map<String,String> sourceOptions = source.getOptions();
+        for (String requiredOption : requiredOptions) {
+            String option = sourceOptions.get(requiredOption);
+            if (option != null) {
+                target.addOption(requiredOption, option);
+            }
+        }
+    }
+
+    protected void driveScanner(Scanner scanner) {
+        for (Map.Entry<Key,Value> entry : scanner) {
+            Result result = new Result(entry.getKey(), entry.getValue());
+
+            boolean offered = false;
+            while (!offered) {
+                try {
+                    offered = resultQueue.offer(result, resultQueueOfferTimeMillis, TimeUnit.MILLISECONDS);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException("Interrupted while offering result", e);
+                }
+            }
+        }
+    }
+
+    private IteratorSetting createScanIterator() {
+        // copy the query data because it is a shared object
+        QueryData queryData = keyWithContext.getContext();
+        IteratorSetting orig = queryData.getSettings().get(0);
+
+        // copy original iterator setting
+        boolean useDocumentIterator = false;
+        IteratorSetting setting;
+        if (useDocumentIterator) {
+            setting = new IteratorSetting(orig.getPriority(), DocumentIterator.class.getSimpleName(), DocumentIterator.class);
+        } else {
+            setting = new IteratorSetting(orig.getPriority(), orig.getName(), orig.getIteratorClass());
+        }
+        setting.addOptions(orig.getOptions());
+
+        // set the query from the query data
+        if (setting.getOptions().containsKey(QueryOptions.QUERY)) {
+            setting.addOption(QueryOptions.QUERY, queryData.getQuery());
+        }
+
+        // serialize query
+        setting.addOption(QueryOptions.SERIAL_EVALUATION_PIPELINE, "true");
+        setting.addOption(QueryOptions.MAX_EVALUATION_PIPELINES, "1");
+        setting.addOption(QueryOptions.MAX_PIPELINE_CACHED_RESULTS, "1");
```

Comment:
I think these are only used by the original iterator? Probably don't need to be set if using the document iterator

Reply:
Correct. Updated the description of https://github.com/NationalSecurityAgency/datawave/issues/2916